### PR TITLE
rust: 1.51.0 -> 1.52.0

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

- update rustc to 1.52.0

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by prusnak** [comment]
@GrahamcOfBorg build rustPackages.rustc rustPackages.rustfmt rustPackages.cargo

**Comment by prusnak** [comment]
I tried to build `fd` and `ripgrep` and they both fail on missing iconv (darwin). Should we just add `libiconv` to `propagatedBuildInputs` of `rustc`?

File: pkgs/build-support/rust/default.nix
Comment (prusnak) [inline pkgs/build-support/rust/default.nix:122]: After the update, I tried to build `fd` and `ripgrep`. They both failed on missing libiconv error. Rather than adding `libiconv` to `buildInputs` of thousands of individual rust packages, I feel this is the better way to go.

Comment (Mic92) [reply pkgs/build-support/rust/default.nix:122]: I tend to agree if this is a dependency of the rust stdlib, otherwise I would keep it rather explicit than bloating the default closure size.. What part in `fd` and `ripgrep` depends on `libiconv`?

Comment (Mic92) [reply pkgs/build-support/rust/default.nix:122]: But let's ask some darwin maintainer: @LnL7 @thefloweringash @abathur

Comment (happysalada) [reply pkgs/build-support/rust/default.nix:122]: I haven't been able to pinpoint it, but I think it's clap.
I've made the fix to about 10 packages so far. I don't mind, the only problem is that for some of them, people don't spot the breakage, until somebody running on darwin comes along.

Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]: Prior discussion: https://github.com/NixOS/nixpkgs/pull/95556

Comment (prusnak) [reply pkgs/build-support/rust/default.nix:122]: I am trying to figure out which fd/ripgrep dependencies require libiconv and trying to fix it via addition to pkgs/build-support/rust/default-crate-overrides.nix but I fail. Any hints on what to do?

Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]: These crate overrides are for `buildRustCrate` (e.g. `crate2nix`) and not `buildRustPackage`.

Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]:
```
$ rg -l "buildRustPackage" | grep "\.nix" |  wc -l                              
442
$ rg -l "buildRustPackage" | grep "\.nix" | xargs -n1 grep -L "libiconv" | wc -l
363
```

And among the delta are probably packages that do not compile because `libiconv` is missing. But I think it's up to @NixOS/darwin-maintainers to make this call.

Comment (siraben) [reply pkgs/build-support/rust/default.nix:122]: @danieldk I have fixed many Rust packages by just adding `libiconv`. I was actually planning to automatically go through Rust packages failing to build because of a lack of it.

I am in favor of making this default on darwin.

Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]: @siraben seems reasonable then. Though I think it would be nice to find the root cause, just so that the addition of this dependency can be properly documented.

I don’t have a Mac, so I have no clue ;).

Comment (abathur) [reply pkgs/build-support/rust/default.nix:122]: I won't pretend to really grok the issues here, but I was looking around for context on a PR about a month ago and found these #37056 #37012

Comment (jonringer) [reply pkgs/build-support/rust/default.nix:122]: I think darwin support can be done in another PR. AFAIK, this would have been an existing issue, and I would rather not delay this PR. Darwin maintainers have around 5 days to come up with a solution before the final staging-next cycle will begin before branch-off of 21.05

Comment (jonringer) [reply pkgs/build-support/rust/default.nix:122]: Also, I don't think having libiconv as a buildInput will bloat too much, as the likelihood of another package requiring should be very high.

**Review by jonringer — APPROVED** [review]
LGTM, thanks :)

```
https://github.com/NixOS/nixpkgs/pull/122110

1 package built:
ripgrep
```

File: pkgs/build-support/rust/default.nix
Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]: I think I have found the culprit. The `libc` crate used to unconditionally link `libiconv` on Apple platforms:

```
    if target.contains("-apple-") {
        println!("cargo:rustc-link-lib=iconv");
    }
```

`libc` is a transitive dependency of *many* widely-used crates, such as `rand`, `memchr`,  and `chrono`, so it percolates through much of the Nix ecosystem.

*However*, this dependency is now made conditional in `libc` (only needed when `libiconv` symbols are used):

https://github.com/rust-lang/libc/pull/2150

So, I propose that we revert adding libiconv as a buildinput, since it will become unnecessary as more and more crates adopt the next version of `libc`.

Comment (prusnak) [reply pkgs/build-support/rust/default.nix:122]: @danieldk Thanks for investigating the issue - I guess we can revert the libiconv addition together with once libc 0.2.95 update once it is out.

Comment (danieldk) [reply pkgs/build-support/rust/default.nix:122]: Proposed PR to remove the dependency #122231.

**Pull request merged**
